### PR TITLE
fix(qwen): batch mixed FlashInfer sampling

### LINKED ISSUE #284: qwen3/qwen35: migrate sampling to the batched FlashInfer primitive
## Problem

The kimi TP1/DP8 sampling work (#237) added a batched sampler (`gpu_sample_batch_into`, `pegainfer-kernels/src/ops/sampling.rs`). The qwen crates still use the legacy single-row path, which has both perf and correctness defects:

**Per-row launches + syncs** — batch decode samples one row at a time:
- `pegainfer-qwen3-4b/src/executor.rs:165` — `for (i, req) in requests` → `sample_from_logits` per row
- `pegainfer-qwen35-4b/src/batch_decode.rs:24` — same loop, plus a per-row `ops::extract_vec` (device alloc + D2D copy per row per step)

Each call is O(1 row): kernels + D2H + `ctx.sync()` → O(batch) syncs per decode step.

**Legacy wrapper defects** (`gpu_sample_flashinfer_cuda`, `pegainfer-kernels/csrc/shared/flashinfer_sampling.cu:125`):
- Softmax is a custom `<<<1, 256>>>` single-block kernel over the 151K vocab — 1 SM busy on a 132-SM GPU. The batched path uses FlashInfer's `OnlineSoftmax` vocab-splitting kernel instead.
- `deterministic=false` (vLLM passes true; batched path passes true).
- Hard-coded 1MB FlashInfer row-state scratch (`FLASHINFER_TOPK_ROW_STATES_BYTES`) — brittle dependency on upstream internal layout.
- Seed is `u64::from(random_val.to_bits())` — a host-drawn f32 used only as seed material, ~30-bit effective space.

## Fix

Adopt the kimi structure: greedy rows keep the batched argmax; non-greedy rows go through one `gpu_sample_batch_into` call per step (3 kernels + 1 sync for the whole batch, per-row temperature/top_k/top_p, fresh u64 seed per step). Then delete the single-row wrapper and its custom softmax.

## Acceptance

- Batch decode with mixed greedy/sampled rows issues one sampling pass per step (verify with nsys: no per-row `logits_to_probs_kernel`).
- `gpu_sample_flashinfer_cuda` and the 1MB row-state scratch are gone.
- hf_golden_gate stays green for both models; greedy TPOT unchanged.

### PR DESCRIPTION
**Pull request by CAICAIIs**
## Summary

Closes #284.

This PR migrates Qwen3/Qwen3.5 mixed greedy and non-greedy token selection to the batched FlashInfer sampling path.

- compact non-greedy rows into one `gpu_sample_batch_into` call per step
- keep greedy and effectively-greedy rows on indexed batched argmax
- replace per-row `random_val` flow with one `sample_seed: u64` per step
- remove the legacy single-row FlashInfer sampler and `logits_to_probs_kernel`
- route Qwen3.5 one-row non-greedy prefill/unified helpers through the batched sampler
- keep Kimi behavior unchanged; only rename the top1 scratch-size helper to make its purpose explicit
- add Qwen3.5 sampling behavior coverage and update the scheduler/kernel docs

## Validation

Static/local checks:

- `cargo fmt --all --check`
- `git diff --check`
- old-symbol scan for `gpu_sample_flashinfer_cuda`, `logits_to_probs_kernel`, `gpu_sample_into`, `FLASHINFER_TOPK_ROW_STATES_BYTES`

GPU validation:

- `cargo check --release --offline -p openinfer-core -p openinfer-kernels`
- `cargo check --release --offline -p openinfer-qwen3-4b`
- `cargo check --release --offline -p openinfer-qwen35-4b --features qwen35-4b`
- `cargo test --release --offline -p openinfer-kernels batch_sampling_top_p_only_small_nucleus_collapses_to_argmax -- --nocapture`
- `cargo test --release --offline -p openinfer-qwen3-4b --tests --no-run`
- `OPENINFER_TEST_MODEL_PATH=<qwen3> cargo test --release --offline -p openinfer-qwen3-4b --test hf_golden_gate -- --nocapture`
- `OPENINFER_TEST_MODEL_PATH=<qwen3> cargo test --release --offline -p openinfer-qwen3-4b --test sampling_behavior -- --nocapture`
- `OPENINFER_TEST_MODEL_PATH=<qwen35> cargo test --release --offline -p openinfer-qwen35-4b --features qwen35-4b --test hf_golden_gate -- --nocapture`
- `OPENINFER_TEST_MODEL_PATH=<qwen35> cargo test --release --offline -p openinfer-qwen35-4b --features qwen35-4b --test sampling_behavior -- --nocapture`
- `OPENINFER_TEST_MODEL_PATH=<qwen35> cargo test --release --offline -p openinfer-qwen35-4b --features qwen35-4b unified_step_decode_matches_graph_decode -- --nocapture`
- `OPENINFER_TEST_MODEL_PATH=<qwen35> cargo test --release --offline -p openinfer-qwen35-4b --features qwen35-4b --test e2e_scheduler -- --nocapture`

## Profiling / Regression Evidence

`nsys profile --trace=cuda,nvtx --cuda-graph-trace=node --export=sqlite` on the Qwen3.5 mixed scheduler test showed the old per-row sampler path is absent:

- `logits_to_probs_kernel`: 0
- `gpu_sample_flashinfer_cuda`: 0
- `gather_cast_logits_f32_kernel`: 100
- `OnlineSoftmaxMapKernel`: 100
- `OnlineSoftmaxReduceKernel`: 100
- `TopKTopPSamplingFromProbKernel`: 100

Qwen3 greedy TPOT A/B, same model and same `request --prompt-len 4096 --output-len 64 --warmup 5 --iters 20 --seed 42` workload:

- upstream/main steady TPOT p50: `6.537992ms`; p99: `6.561389ms`; output hash: `83f4c3f2614d57b5`
- patched steady TPOT p50: `6.508904ms`; p99: `6.540855ms`; output hash: `83f4c3f2614d57b5`
- p50 delta: `-0.44%`

## Notes

This changes non-greedy sampling seed plumbing, so sampled token streams are not expected to be bit-compatible with the old per-row path.

This PR does not claim HTTP serving benchmark coverage. That follow-up is tracked in #412.

Related: #393

### CODE REVIEW

**Review by xiaguan — APPROVED** [review]
Reviewed on a 5090 (sm_120). Core design is solid and the real correctness gates pass:

- **Routing** (`select_batch_tokens_into`): greedy + effectively-greedy (`top_p <= 1/vocab`) → indexed batched argmax; non-greedy compacted into one `gpu_sample_batch_into`; tokens land back at their original row indices.
- **Seed**: one `StdRng::seed_from_u64(seed)` per scheduler, advanced to a fresh `sample_seed: u64` per step — one engine seed, per-step advance, no per-request seed. 👍
- **`flashinfer_top1_row_states_bytes` fix (`73fd94a`)**: sizing the multi-CTA deterministic radix-collect scratch by SM count with a 1 MiB floor is a real correctness fix — the old `sizeof(RadixRowState)` under-allocated.
- **qwen35 `sampling_behavior.rs`**: nice E2E behavioral gate (greedy determinism, `top_k=1` / tiny `top_p` collapse, high-temp variation across 4 runs); guards the #284 silent-greedy regression.

Verified on qwen3 weights: `hf_golden_gate` ✅ (41.9s), `sampling_behavior` ✅ (8.7s).

**One ask before merge: please drop (or tolerance-ize) `batch_matches_sequential`.** It still asserts exact `batch == sequential` token equality and **fails on a 5090**:

```
batch_decode::tests::batch_matches_sequential ... FAILED
batch_decode.rs:859: assertion `left == right` failed: Decode request A mismatch
```

This is the batch-invariance issue I just filed as #414: batch composition changes the GEMM reduction order and drifts logits ~1 ULP, which flips a near-tie argmax and diverges the two runs. It's hardware-dependent (passes on your GPU, fails on the 5090) and model-gated, so it silently rots in CI. `hf_golden_gate`'s tolerance check already covers batched decode (it passes here).

Otherwise LGTM for #284. 🚀
